README.md
Telescope is a simple window/task switcher for linux that, when activated, pops up live displays of all windows and lets you switch to or close any of them.

The original project was located at https://bitbucket.org/IlyaSkriblovsky/telescope but hasn't been updated in 9 months and used mercurial. I'm making changes to make it work better with archlinux and dwm, but nearly all the code is from llyaSkriblovsky on bitbucket.


debian/changelog
telescope (2.0~rc14) unstable; urgency=low

  * Fixed bug: Telescope may crash when .desktop file changed

 -- Ilya Skriblovsky <Ilya.Skriblovsky@gmail.com>  28 Jun 2011 23:50:00 +0300

telescope (2.0~rc13) unstable; urgency=low

  * Fixed bug with empty applications.menu

 -- Ilya Skriblovsky <Ilya.Skriblovsky@gmail.com>  27 Jun 2011 22:20:00 +0300

telescope (2.0~rc12) unstable; urgency=low

  * Fixed bug that made telescope crashing if ~/.osso/menus/applications.menu is missing. Now /etc/xdg/menus/applications.menu will be used in this case.

  * New icon from Kroll!

 -- Ilya Skriblovsky <Ilya.Skriblovsky@gmail.com>  25 Apr 2011 16:39:00 +0300

telescope (2.0~rc11) unstable; urgency=low

  * Now user can select icons for categories using longtap

 -- Ilya Skriblovsky <Ilya.Skriblovsky@gmail.com>  1 Mar 2011 14:36:00 +0300

telescope (2.0~rc10) unstable; urgency=low

  * Fixed Launcher panel in portrait mode

 -- Ilya Skriblovsky <Ilya.Skriblovsky@gmail.com>  26 Feb 2011 14:52:00 +0300

telescope (2.0~rc9) unstable; urgency=low

  * Internal refactoring of image handling code

  * New appearance of bottom bar in Launcher

 -- Ilya Skriblovsky <Ilya.Skriblovsky@gmail.com>  25 Feb 2011 11:57:00 +0300

telescope (2.0~rc8) unstable; urgency=low

  * Left-right buttons in Launcher were replaced by bottom bar with buttons for each section

 -- Ilya Skriblovsky <Ilya.Skriblovsky@gmail.com>  23 Feb 2011 22:05:00 +0300

telescope (2.0~rc7) unstable; urgency=low

  * Fixed zombie processes

  * Added globalpress keymapping event

 -- Ilya Skriblovsky <Ilya.Skriblovsky@gmail.com>  22 Feb 2011 23:07:00 +0300

telescope (2.0~rc6) unstable; urgency=low

  * Fixed synchronization of DBus and Xlib threads. Now DBus thread doesn't show windows by itself, but sends ClientMessage to X thread instead

 -- Ilya Skriblovsky <Ilya.Skriblovsky@gmail.com>  8 Oct 2010 13:56:00 +0400

telescope (2.0~rc5) unstable; urgency=low

  * Sections now draw itself on the fly without creating offscreen pixmap for each section. Instead, offscreen pixmaps are created for every icon.

  * Portrait mode support for Launcher

 -- Ilya Skriblovsky <Ilya.Skriblovsky@gmail>  1 Oct 2010 16:16:00 +0400

telescope (2.0~rc4) unstable; urgency=low

  * Fixed bug: When SystemUI is disabled Launcher window doesn't hides properly

  * Reduced flickering while switching between Telescope and Launcher with hot key

 -- Ilya Skriblovsky <Ilya.Skriblovsky@gmail.com>  17 Sep 2010 11:33:00 +0300

telescope (2.0~rc3) unstable; urgency=low

  * launcher.disable option in telescope.conf for disabling app launcher

 -- Ilya Skriblovsky <Ilya.Skriblovsky@gmail.com>  11 Sep 2010 13:08:00 +0300

telescope (2.0~rc2) unstable; urgency=low

  * XftFontMatch & XftDestroyPattern replaced with XftFontOpen, fixing some strange segfaults

 -- Ilya Skriblovsky <Ilya.Skriblovsky@gmail.com>  09 Aug 2010 15:19:00 +0300

telescope (2.0~rc1) unstable; urgency=low

  * Telescope merged with Launcher from Andry Gunawan

 -- Ilya Skriblovsky <Ilya.Skriblovsky@gmail.com>  07 Aug 2010 23:08:00 +0300

telescope (1.0.3) unstable; urgency=low

  * Fix with screen rotation

 -- Ilya Skriblovsky <Ilya.Skriblovsky@gmail.com>  27 Jul 2010 21:20:00 +0300

telescope (1.0.2) unstable; urgency=low

  * In rare cases when telescope is unable to grab a screenshot, it will draw gray checkers instead of it

  * Now telescope can not be called when no windows opened (strange, but some users don't like to see their nice wallpapers)

  * XTools now can extract window's icon from _NET_WM_ICON propery, but this functionality is not used

 -- Ilya Skriblovsky <Ilya.Skriblovsky@gmail.com>  27 Jul 2010 15:22:00 +0300

telescope (1.0.1) stable; urgency=low

  * Added install notice about Home key

 -- Ilya Skriblovsky <Ilya.Skriblovsky@gmail.com>  14 Jul 2010 23:35:00 +0300

telescope (1.0) stable; urgency=low

  * The copy of 0.2.9

 -- Ilya Skriblovsky <Ilya.Skriblovsky@gmail.com>  14 Jul 2010 13:55:00 +0300

telescope (0.2.9) unstable; urgency=low

  * D-Bus support. Telescope now can be called by calling org.telescope bus
    with path /Telescope and method org.telescope.Telescope.Show (and Hide)

  * Workaround for liqbase apps. They will be iconified before showing
    Telescope

 -- Ilya Skriblovsky <Ilya.Skriblovsky@gmail.com>  09 Jul 2010 15:50:00 +0300

telescope (0.2.8.1) unstable; urgency=low

  * Telescope activating key can be changed via telescope.conf

  * Little change in how Telescope grabs its activating key

 -- Ilya Skriblovsky <Ilya.Skriblovsky@gmail.com>  06 Jul 2010 14:55:00 +0300

telescope (0.2.8) unstable; urgency=low

  * Huge code refactoring without visible changes for user

  * show.desktop.thumbnail and show.desktop.iconify config options
    for Xlab

 -- Ilya Skriblovsky <Ilya.Skriblovsky@gmail.com>  29 Jun 2010 17:30:00 +0300

telescope (0.2.7.3) unstable; urgency=low

  * Another fix for osso mediaplayer

  * Keybindings are loaded at first use (hack for maemo)

 -- Ilya Skriblovsky <Ilya.Skriblovsky@gmail.com>  07 Jun 2010 16:55:00 +0300

telescope (0.2.7.2) unstable; urgency=low

  * Fix for osso mediaplayer

 -- Ilya Skriblovsky <Ilya.Skriblovsky@gmail.com>  02 Jun 2010 17:15:00 +0300

telescope (0.2.7.1) unstable; urgency=low

  * Fixed #5771 "Telescope does not works with plain color backgrounds"

  * 'selection.disabled' option for disabling all window selection stuff

 -- Ilya Skriblovsky <Ilya.Skriblovsky>  01 Jun 2010 09:00:00 +0300

telescope (0.2.7) unstable; urgency=low

  * Fixed handling of non-conformant windows, for example, SDL games

 -- Ilya Skriblovsky <Ilya.Skriblovsky>  31 May 2010 16:00:00 +0300

telescope (0.2.6.2) unstable; urgency=low

  * Fixed #5779 "Window titles are not displaying sometimes"

  * Fixed #5786 "Window previews are sometimes invalid"

  * Fixed #5784 "Active window is not highlighted sometimes"

 -- Ilya Skriblovsky <Ilya.Skriblovsky>  22 May 2010 22:40:00 +0300

telescope (0.2.6.1) unstable; urgency=low


DBus.cpp
//
// Telescope - graphical task switcher
//
// (c) Ilya Skriblovsky, 2010
// <Ilya.Skriblovsky@gmail.com>
//

// $Id: DBus.cpp 174 2011-02-28 17:17:05Z mitrandir $

#ifdef DBUS

#include "DBus.h"

#include "XEventLoop.h"
#include "TeleWindow.h"
#include "LauncherWindow.h"

DBus * DBus::_instance = 0;


#ifdef LAUNCHER
DBus::DBus(XEventLoop *eventLoop, TeleWindow *teleWindow, LauncherWindow *launcherWindow)
#else
DBus::DBus(XEventLoop *eventLoop, TeleWindow *teleWindow)
#endif
{
    _instance = this;

    _teleWindow = teleWindow;

    #ifdef LAUNCHER
        _launcherWindow = launcherWindow;
    #endif


    DBusError error;
    dbus_error_init(&error);

    _conn = dbus_bus_get(DBUS_BUS_SESSION, &error);
    if (dbus_error_is_set(&error) || _conn == 0)
    {
        fprintf(stderr, "D-Bus connection error: %s\n", error.message);
        return;
    }

    _dispatchOnIdle = true;
    dbus_connection_set_dispatch_status_function(
        _conn,
        dispatchStatusChanged,
        this,
        0
    );

    dbus_bus_request_name(_conn, "org.telescope",
            DBUS_NAME_FLAG_DO_NOT_QUEUE, &error);
    if (dbus_error_is_set(&error))
    {
        fprintf(stderr, "D-Bus bus name error: %s\n", error.message);
        return;
    }


    _telescopeVTable.unregister_function = (DBusObjectPathUnregisterFunction)telescope_unregister;
    _telescopeVTable.message_function = (DBusObjectPathMessageFunction)telescope_message;
    dbus_connection_register_object_path(_conn, "/Telescope",
        &_telescopeVTable, this);


    #ifdef LAUNCHER
        _launcherVTable.unregister_function = (DBusObjectPathUnregisterFunction)launcher_unregister;
        _launcherVTable.message_function = (DBusObjectPathMessageFunction)launcher_message;
        dbus_connection_register_object_path(_conn, "/Launcher",
            &_launcherVTable, this);
    #endif



    eventLoop->addIdleTask(this);
    eventLoop->addDBusConnection(_conn);
}

DBus::~DBus()
{
    dbus_connection_unref(_conn);
}


void DBus::dispatchStatusChanged(DBusConnection *conn, DBusDispatchStatus status, void *data)
{
    if (status == DBUS_DISPATCH_DATA_REMAINS)
        static_cast<DBus*>(data)->_dispatchOnIdle = true;
}


void DBus::onIdle()
{
    if (_dispatchOnIdle)
    {
        _dispatchOnIdle = false;

        while (dbus_connection_get_dispatch_status(_conn) == DBUS_DISPATCH_DATA_REMAINS)
            dbus_connection_dispatch(_conn);
    }
}







void DBus::telescope_unregister(DBusConnection *conn, DBus *self)
{
}

DBusHandlerResult DBus::telescope_message(
    DBusConnection *conn,
    DBusMessage *msg,
    DBus *self
)
{
    if (dbus_message_is_method_call(msg, "org.telescope.Telescope", "Show"))
    {
        DBusMessage *reply = dbus_message_new_method_return(msg);
        dbus_connection_send(conn, reply, 0);
        dbus_message_unref(reply);

        self->_teleWindow->show();
    }
    else if (dbus_message_is_method_call(msg, "org.telescope.Telescope", "Hide"))
    {
        DBusMessage *reply = dbus_message_new_method_return(msg);
        dbus_connection_send(conn, reply, 0);
        dbus_message_unref(reply);

        self->_teleWindow->hide();
    }

    XFlush(self->_teleWindow->display());

    return DBUS_HANDLER_RESULT_HANDLED;
}


#ifdef LAUNCHER

void DBus::launcher_unregister(DBusConnection *conn, DBus *self)
{
}

DBusHandlerResult DBus::launcher_message(
    DBusConnection *conn,
    DBusMessage *msg,
    DBus *self
)
{
    if (dbus_message_is_method_call(msg, "org.telescope.Launcher", "Show"))
    {
        DBusMessage *reply = dbus_message_new_method_return(msg);
        dbus_connection_send(conn, reply, 0);
        dbus_message_unref(reply);

        self->_launcherWindow->show();
    }
    else if (dbus_message_is_method_call(msg, "org.telescope.Launcher", "Hide"))
    {
        DBusMessage *reply = dbus_message_new_method_return(msg);
        dbus_connection_send(conn, reply, 0);
        dbus_message_unref(reply);

        self->_launcherWindow->hide();
    }

    XFlush(self->_launcherWindow->display());

    return DBUS_HANDLER_RESULT_HANDLED;
}

#endif


#endif
